crates/oxc_linter/src/rules/eslint/no_useless_call.rs

```diff
@@ -0,0 +1,184 @@
+);
+
+fn no_useless_call_diagnostic(span1: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Disallow the use of undeclared variables.").with_label(span1)
+}
+
+fn is_array_argument(expr: Option<&Argument>) -> bool {
+    matches!(expr, Some(Argument::ArrayExpression(_)))
+}
+
+fn is_call_or_non_variadic_apply(call_expr: &CallExpression) -> bool {
+    let Some(member_expr) = skip_chain_expression(&call_expr.callee) else {
+        return false;
+    };
+
+    let Some(static_name) = member_expr.static_property_name() else {
+        return false;
+    };
+
+    (static_name == "call" && call_expr.arguments.len() >= 1)
+        || (static_name == "apply"
+            && call_expr.arguments.len() == 2
+            && is_array_argument(call_expr.arguments.get(1)))
+}
+
+fn is_validate_this_arg(
+    ctx: &LintContext,
+    expected_this: Option<&Expression>,
+    this_arg: &Expression,
+) -> bool {
+    match expected_this {
+        Some(expected_this) => is_same_reference(
+            expected_this.get_inner_expression(),
+            this_arg.get_inner_expression(),
+            ctx,
+        ),
+        None => this_arg.is_null_or_undefined(),
+    }
+}
+
+impl Rule for NoUselessCall {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else {
+            return;
+        };
+
+        if !is_call_or_non_variadic_apply(call_expr) {
+            return;
+        }
+        let Some(callee_member) = skip_chain_expression(&call_expr.callee) else {
+            return;
+        };
+
+        let expected_this =
+            skip_chain_expression(callee_member.object()).map(|member| member.object());
+
+        let Some(this_arg) = call_expr.arguments.first() else {
+            return;
+        };
+        let Some(this_expr) = this_arg.as_expression() else {
+            return;
+        };
+
+        if is_validate_this_arg(ctx, expected_this, this_expr) {
+            ctx.diagnostic(no_useless_call_diagnostic(call_expr.callee.span()));
+        }
+    }
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+
+    let pass = vec![
+        "foo.apply(obj, 1, 2);",
+        "obj.foo.apply(null, 1, 2);",
+        "obj.foo.apply(otherObj, 1, 2);",
+        "a.b(x, y).c.foo.apply(a.b(x, z).c, 1, 2);",
+        "foo.apply(obj, [1, 2]);",
+        "obj.foo.apply(null, [1, 2]);",
+        "obj.foo.apply(otherObj, [1, 2]);",
+        "a.b(x, y).c.foo.apply(a.b(x, z).c, [1, 2]);",
+        "a.b.foo.apply(a.b.c, [1, 2]);",
+        "foo.apply(null, args);",
+        "obj.foo.apply(obj, args);",
+        "var call; foo[call](null, 1, 2);",
+        "var apply; foo[apply](null, [1, 2]);",
+        "foo.call();",
+        "obj.foo.call();",
+        "foo.apply();",
+        "obj.foo.apply();",
+        "obj?.foo.bar.call(obj.foo, 1, 2);",
+        "class C { #call; wrap(foo) { foo.#call(undefined, 1, 2); } }",
+        //"(obj?.foo).test.bar.call(obj?.foo.test, 1, 2);",
```

Comment:
eslint ignores this case because of the parenthesis on the `(obj?.foo)`.

i think we should report this